Source/DafnyRuntime/DafnyRuntimeRust/src/lib.rs

```diff
@@ -9,6 +16,346 @@ pub use num::NumCast;
+        length: SizeT,
+    },
+    LazySequence {
+        is_string: bool,
+        node_count: usize,
+        length: SizeT,
+        boxed: RefCell<Rc<Sequence<T>>>
+    }
+}
+
+#[allow(dead_code)]
+impl <T> Sequence<T>
+where T: Clone {
+    fn is_string(&self) -> bool {
+        match self {
+            Sequence::ArraySequence { is_string, .. } => *is_string,
+            Sequence::ConcatSequence { is_string, .. } => *is_string,
+            Sequence::LazySequence { is_string, .. } => *is_string,
+        }
+    }
+    fn new_array_sequence(values: &Rc<Vec<T>>) -> Rc<Sequence<T>> {        
+        Sequence::<T>::new_array_sequence_is_string(values, false)
+    }
+    fn new_array_sequence_is_string(values: &Rc<Vec<T>>, is_string: bool) -> Rc<Sequence<T>> {        
+        Rc::new(Sequence::ArraySequence {
+            is_string,
+            node_count: 1,
+            values: Rc::clone(values),
+        })
+    }
+    fn new_concat_sequence(left: &Rc<Sequence<T>>, right: &Rc<Sequence<T>>) -> Rc<Sequence<T>> {
+        Sequence::<T>::new_concat_sequence_is_string(left, right, false)
+    }
+    fn new_concat_sequence_is_string(left: &Rc<Sequence<T>>, right: &Rc<Sequence<T>>, is_string: bool) -> Rc<Sequence<T>> {
+        Rc::new(Sequence::ConcatSequence {
+            is_string: is_string || (left.is_string() && right.is_string()),
+            node_count: 1 + left.node_count() + right.node_count(),
+            left: Rc::clone(&left),
+            right: Rc::clone(&right),
+            length: left.cardinality() + right.cardinality(),
+        })
+    }
+    fn new_lazy_sequence(boxed: &Rc<Sequence<T>>) -> Rc<Sequence<T>> {
+        Sequence::<T>::new_lazy_sequence_is_string(boxed, false)
+    }
+    fn new_lazy_sequence_is_string(underlying: &Rc<Sequence<T>>, is_string: bool) -> Rc<Sequence<T>> {
+        Rc::new(Sequence::LazySequence {
+            is_string: is_string || underlying.is_string(),
+            node_count: underlying.node_count() + 1,
+            length: underlying.cardinality(),
+            boxed: RefCell::new(Rc::clone(underlying)),
+        })
+    }
+
+    fn to_array(&self) -> Rc<Vec<T>> {
+        // We convert the match above to statements using the Rust "it" idiom
+        if let Sequence::ArraySequence { values, .. } = self {
+            // The length of the elements
+            Rc::clone(values)
+        } else if let Sequence::ConcatSequence { length, .. } = self {
+            let mut array: Vec<T> = Vec::with_capacity(*length);
+            Sequence::<T>::append_recursive(&mut array, self);
+            Rc::new(array)
+        } else if let Sequence::LazySequence { boxed, is_string, ..  } = self {
+            let result = boxed.borrow().to_array();
+            // Put the value back into boxed
+            boxed.replace(Rc::clone(
+                &Sequence::<T>::new_array_sequence_is_string(&result, *is_string)));
+            result
+        } else {
+            panic!("This should never happen")
```

Comment:
Why is this necessary here but not in the `match this` cases?

Reply:
Because Rust's if sentences are like if-then-else in Dafny, they are types, so they always require an else (there is no return keyword, only the last value is returned). panic returning the empty type, it type checks.

In the newest PR, this if then else is transformed back to a match, as I'm getting more used to Rust's syntax I was able to make it work.